docs: add developing locally doc to the guide (#349)
screwdriver-cd/screwdriver-cd.github.io: docs/cluster-management/developing-locally.md

---
layout: main
title: Configuring the UI
category: Cluster Management
menu: menu
toc:
    - title: Developing locally
      url: "#developing-locally"
      active: true
    - title: Developing locally with executor-queue
      url: "#developing-locally-with-executor-queue"

---

> ## Developing Locally

Reviewer: I think it would be good to have a sentence or two highlighting the intent of this doc vs https://docs.screwdriver.cd/cluster-management/running-locally

And it might be good to add a link or something to your new doc from this one: https://docs.screwdriver.cd/about/contributing/getting-started-developing

---
layout: main
title: Configuring the UI
category: Cluster Management
menu: menu
toc:
    - title: Developing locally
      url: "#developing-locally"
      active: true
    - title: Developing locally with executor-queue
      url: "#developing-locally-with-executor-queue"

---
## Developing Locally

### Step 1: Map domain name sd.screwdriver.cd to your ip in hosts file
* Append this line to your /etc/hosts file:

  ```
  127.0.0.1 sd.screwdriver.cd
  ```

### Step 2: Create a new Github OAuth Application
Go to Settings > Developer settings > OAuth Apps, click `New OAuth App` button and configure as described below:

* Application Name: (choose for yourself)
* Homepage URL: `http://sd.screwdriver.cd:4200`
* Application description: (choose for yourself)
* Authorization callback URL: `http://sd.screwdriver.cd:9001/v4/auth/login`

> See screenshot as below

Reviewer: ```suggestion
See screenshot below
```

![developing-locally-ouath](./assets/developing-locally-ouath.png)

> > Take note of the client ID and the client Secret, you'll need them in following steps

Reviewer: ```suggestion
> Take note of the client ID and the client Secret, you'll need them in the following step
```

## Step 3: Clone necessary repos from github of screwdriver-cd organization:
* [ui](https://github.com/screwdriver-cd/ui)
* [screwdriver](https://github.com/screwdriver-cd/screwdriver)
* [store](https://github.com/screwdriver-cd/store)

```bash
git clone git@github.com:screwdriver-cd/ui.git
git clone git@github.com:screwdriver-cd/screwdriver.git
git clone git@github.com:screwdriver-cd/store.git
```

## Step 4: Add local config files for these three repos
Create a file called `local.js` in `ui/config` and `local.yaml` in `screwdriver/config` and `store/config` folders.

### ui/config/local.js

```javascript
let SDAPI_HOSTNAME;
let SDSTORE_HOSTNAME;

SDAPI_HOSTNAME = 'http://sd.screwdriver.cd:9001';
SDSTORE_HOSTNAME = 'http://sd.screwdriver.cd:9002';

module.exports = {
  SDAPI_HOSTNAME,
  SDSTORE_HOSTNAME
};
```

### screwdriver/config/local.yaml

> * Remember to fill in your Github OAuth **client id** and OAuth **client secret**, you can find them in the OAuth application you created in Step2

Reviewer: ```suggestion
* Fill in your Github OAuth **client id** (oauthClientId) and OAuth **client secret**, (oauthClientSecret) you can find them in the OAuth application you created in Step2
```

> * Remember to generate your own **jwtPrivateKey** and **jwtPublicKey** using

Reviewer: ```suggestion
* Generate your own **jwtPrivateKey** (jwtPrivateKey) and **jwtPublicKey** (jwtPublicKey) using
```

    ```bash
    openssl genrsa -out jwt.pem 2048
    openssl rsa -in jwt.pem -pubout -out jwt.pub
    ```

> * Remember to create a folder called "mw-data": `mkdir mw-data`

Reviewer: ```suggestion
* Create a folder called "mw-data" using `mkdir mw-data` in your screwdriver repo
```

> * Look up your ip first: `ifconfig`, YOUR_IP

Reviewer: ```suggestion
* Fill in your ip (YOUR_IP), look up your ip first using `ifconfig`
```

  host: 0.0.0.0

  # Externally routable URI (usually your load balancer or CNAME)
  # This requires to be a routable IP inside docker for executor, see
  # https://github.com/screwdriver-cd/screwdriver/blob/095eaf03e053991443abcbde91c62cfe06a28cba/lib/server.js#L141
  uri: http://YOUR_IP:9001

ecosystem:
  # Externally routable URL for the User Interface
  ui: http://sd.screwdriver.cd:4200

  # Externally routable URL for the Artifact Store
  store: http://YOUR_IP:9002

  allowCors: ['http://sd.screwdriver.cd', 'http://YOUR_IP:9001']

executor:
    plugin: docker
    docker:
        enabled: true
        options:
            docker:
                socketPath: "/var/run/docker.sock"

scms:
    github:
        plugin: github
        config:
            # github
            oauthClientId: your-oauth-client-id
            oauthClientSecret: your-oauth-client-secret
            secret: a-really-real-secret
            username: sd-buildbot
            email: dev-null@screwdriver.cd

datastore:
  plugin: sequelize
  sequelize:
    # Type of server to talk to
    dialect: sqlite
    # Storage location for sqlite
    storage: ./mw-data/storage.db
```

### store/config/local.yaml

> * Similarly like the `mw-data`, remember to create a folder called "store-data": `mkdir store-data`

Reviewer: ```suggestion
* Similar to the `mw-data` file for the screwdriver repo, you will need to create a folder called "store-data" in your store repo using `mkdir store-data`
```

```
auth:
    # A public key for verifying JWTs signed by api.screwdriver.cd
    jwtPublicKey: |
        -----BEGIN PUBLIC KEY-----
        ******SOME KEYS HERE******
        -----END PUBLIC KEY-----

strategy:
    plugin: disk
    disk:
        cachePath: './store-data'
        cleanEvery: 3600000
        partition : 'cache'

httpd:
    port: 9002

ecosystem:
    # Externally routable URL for the User Interface
    ui: http://sd.screwdriver.cd:4200

    # Externally routable URL for the Artifact Store
    api: http://sd.screwdriver.cd:9001

    allowCors: ['http://sd.screwdriver.cd']
```

### Step 5: Install dependencies and you are ready to go!
Just need to run below commands inside each repo

```
npm install && npm run start
```

> While all ui, screwdriver and store are runing, now you can go to `http://sd.screwdriver.cd:4200` to check out.

Reviewer: ```suggestion
While the UI, Screwdriver API, and Store apps are running, you can visit `http://sd.screwdriver.cd:4200` in your browser to interact with your local Screwdriver.
```

## Developing locally with executor-queue

> > TODO

Reviewer 2: perhaps @pritamstyz4ever can chime in here ? Haven't you create a queue based setup for build timeout work ?

    # Storage location for sqlite
    storage: ./mw-data/storage.db
```

### store/config/local.yaml
* Similarly like the `mw-data`, remember to create a folder called "store-data": `mkdir store-data`

```
auth:
    # A public key for verifying JWTs signed by api.screwdriver.cd
    jwtPublicKey: |
        -----BEGIN PUBLIC KEY-----
        ******SOME KEYS HERE******
        -----END PUBLIC KEY-----

strategy:
    plugin: disk
    disk:
        cachePath: './store-data'
        cleanEvery: 3600000
        partition : 'cache'

httpd:
    port: 9002

ecosystem:
    # Externally routable URL for the User Interface
    ui: http://sd.screwdriver.cd:4200

    # Externally routable URL for the Artifact Store
    api: http://sd.screwdriver.cd:9001

    allowCors: ['http://sd.screwdriver.cd']
```

### Step 5: Install dependencies and you are ready to go!
Just need to run below commands inside each repo

```
npm install && npm run start
```

While all ui, screwdriver and store are runing, now you can go to `http://sd.screwdriver.cd:4200` to check out.

## Developing locally with executor-queue

### Step 1: Install redis server and client
```
brew update
brew install redis
```

> To have launchd start redis now and restart at login:

Reviewer: ```suggestion
To have launchd, start Redis now and restart at login:
```

```
brew services start redis
```

Or, if you don't want/need a background service you can just run:

```
redis-server /usr/local/etc/redis.conf
```

> Test if Redis server is running.

Reviewer: ```suggestion
Test to see if the Redis server is running.
```

```
redis-cli ping
```
If it replies “PONG”, then it’s good to go!

> Location of Redis configuration file. Modfy "requirepass" if you want to set password.

Reviewer: ```suggestion
Location of the Redis configuration file. Modify "requirepass" if you want to set the password.
```

```
/usr/local/etc/redis.conf
```

Uninstall Redis and its files.

```
brew uninstall redis
rm ~/Library/LaunchAgents/homebrew.mxcl.redis.plist
```
### Step 2: Modify screwdriver/config/local.yaml and change executor configuration
```
executor:
    plugin: queue
    queue:
        enabled: true
        options:
            # Configuration of the redis instance containing resque
            redisConnection:
                host: "127.0.0.1"
                port: 6379
                options:
                    # Set password if set in previous step
                    password: 'a-secure-password-not-welcome123'
                    tls: false
                database: 0
```

### Step 3: Clone repository [queue-worker](https://github.com/screwdriver-cd/queue-worker) and modify default.yaml
```
git clone git@github.com:screwdriver-cd/queue-worker.git
```
### queue-worker/config/default.yaml
``` 
 executor:
    plugin: docker
    docker:
      enabled: true
      options:
        docker:
            socketPath: "/var/run/docker.sock"
           
 ecosystem:
    # Externally routable URL for the User Interface

>     ui: http://sd.ouroath.com:4200

Author: We need to change this to `sd.screwdriver.cd`
